### ISSUE TITLE
Redirect with base?

### ISSUE THREAD

**Issue by micmath** [mikecao/flight#37]
I've been scratching my head about why Flight has the following code in it's source:

https://github.com/mikecao/flight/blob/master/flight/Flight.php#L367

The problem I'm having with it, as it is written, is that my site is based on a subdomain/rewrite system. So, if you go to `http://test.example.com/index.php`, this hits a rewrite rule that shows the contents of `public_html/test/index.php`, where "public_html" is my document root.

In this state the value of Flight's "base" is set to `dirname(getenv('SCRIPT_NAME'))`, which is "/test" on my server. Note however that, due to the rewrite, the url in the browser does not contain the "/test" fragment, instead it just reads: "http://test.example.com/index.php".

Now the part I don't understand: if I then execute `Flight::redirect('/login', 303)`, Flight's code I linked to above will prepend the $base to the url I am redirecting to, turning the correct URL of `/login` into the incorrect `/test/login`. But what is the reason it should do that? I could understand if I were redirecting to `Flight::redirect('login', 303)` (note I've removed the leading slash). In that case, yes you would need to try to add some filepath information before the plain string "login". But in my case, I've already explicitly written the full path that I want to redirect to: `/login`. This is already usable as the value of a Redirect header, and doesn't need any `$base` to be prepended.

So, in the case that the `$url` being redirected to begins with "/" (or contains "://") shouldn't Flight just use the value as given, and not try to add a `$base`?

**Comment by mikecao** [comment]
The idea behind the base is so Flight can run from a subdirectory. If running from the root, the base is `/`, while under a subdirectory it would be `/subdir`. So when you go to redirect, you simply need to provide the route you're going to without worrying about the structure. Flight will either use `/login` or `/subdir/login` depending on the base. Flight treats all urls relative to itself, which or may not be a good thing, I'll have to think about it. It currently does handle `://` correctly though.

However, when you're rewriting urls like in your case, it's not working as expected. I'll have to run some tests and update how Flight handles the base, possibly making it a config.

**Comment by micmath** [comment]
Thanks to the awesome extensibility of Flight, I can solve my specific problem, by adding the following to my `index.php`, like so:

``` php
$base = dirname(getenv('SCRIPT_NAME'));
Flight::map('redirect', function($url, $code = 303) use ($base) {
    if ($base != '/' && strpos($url, '://') === false && $url[0] !== '/') {
        $url = $base.'/'.$url;
    }

    Flight::response(false)
        ->status($code)
        ->header('Location', $url)
        ->write($url)
        ->send();
});
```

If including this in the core would break some other usecase, then I agree, having the ability to configure the value of `$base` would be a good alternative, and would mean I wouldn't have to override the entire `redirect` just to get the base prepending to do what I want.

**Comment by micmath** [comment]
Possibly the pattern to follow is the way CodeIgniter allows you to set `$config['base_url']`, but will try to calculate it for you if you leave it unconfigured.

**Comment by micmath** [comment]
Re "I'll have to run some tests and update how Flight handles the base," any update?

**Comment by mikecao** [comment]
I'm going to add a configuration setting called `flight.base_url` which will let you override the derived value. Need to run some tests first.

**Comment by mikecao** [comment]
Now available in commit d41795cbdc72b88d1a8e6478b719e053d8c85384.

**Comment by ncie66** [comment]
p
